core/GlobalState.h

```diff
@@ -40,6 +42,147 @@ class Serializer;
+                                    uint32_t num) {
+            return [&gs, uniqueNameKind, original, num](auto name) {
+                if (name.kind() != NameKind::UNIQUE) {
+                    return false;
+                }
+
+                auto data = name.dataUnique(gs);
+                return data->uniqueNameKind == uniqueNameKind && data->num == num && data->original == original;
+            };
+        }
+    };
+
+private:
+    std::vector<Bucket> buckets_;
+
+    void moveNames(Bucket *from, Bucket *to, unsigned int szFrom, unsigned int szTo);
+
+public:
+    static inline Hash hashMixUTF8(std::string_view name) {
+        return _hash(name);
+    }
+
+    static inline Hash hashMixUnique(UniqueNameKind unk, unsigned int num, unsigned int rawId) {
+        return mix(mix(num, static_cast<uint32_t>(unk)), rawId) * HASH_MULT2 + static_cast<uint32_t>(NameKind::UNIQUE);
+    }
+
+    static inline Hash hashMixConstant(unsigned int id) {
+        return id * HASH_MULT2 + static_cast<uint32_t>(NameKind::CONSTANT);
+    }
+
+    static inline Hash hashNameRef(const GlobalState &gs, NameRef nref) {
+        switch (nref.kind()) {
+            case NameKind::UTF8:
+                return _hash(nref.shortName(gs));
+            case NameKind::CONSTANT:
+                return hashMixConstant(nref.dataCnst(gs)->original.rawId());
+            case NameKind::UNIQUE: {
+                auto data = nref.dataUnique(gs);
+                return hashMixUnique(data->uniqueNameKind, data->num, data->original.rawId());
+            }
+        }
+    }
+
+    bool empty() const {
+        return this->buckets_.empty();
+    }
+
+    size_t size() const {
+        return this->buckets_.size();
+    }
+
+    size_t capacity() const {
+        return this->buckets_.capacity();
+    }
+
+    void clear() {
+        this->buckets_.clear();
+    }
+
+    void resize(size_t size) {
+        ENFORCE_NO_TIMER((size & (size - 1)) == 0, "namesByHashSize is not a power of 2");
+        this->buckets_.resize(size);
+    }
+
+    void reserve(size_t size) {
+        ENFORCE_NO_TIMER((size & (size - 1)) == 0, "namesByHashSize is not a power of 2");
+        this->buckets_.reserve(size);
+    }
+
+    absl::Span<const Bucket> buckets() const {
+        return absl::MakeSpan(this->buckets_);
+    }
+
+    void expandNames(uint32_t utf8NameSize, uint32_t constantNameSize, uint32_t uniqueNameSize);
+
+    template <typename Pred> const Bucket &lookupBucket(Hash hash, Pred &&pred) const {
+        unsigned int hashTableSize = buckets_.size();
+        unsigned int mask = hashTableSize - 1;
+        auto bucketId = hash & mask;
+        unsigned int probeCount = 1;
+
+        while (buckets_[bucketId].present()) {
+            auto &bucket = buckets_[bucketId];
+            if (bucket.hash == hash) {
+                auto name = NameRef::fromRawUnchecked(bucket.rawId);
+                if (std::invoke(pred, name)) {
```

Comment:
This is a very nice way of splitting out the logic here for the reading and the writing paths.